### PR TITLE
allow for link-based session locks on desktop

### LINKED ISSUE #3521: Cannot save file after New Session, file on network drive
### System details

    RStudio Edition : Desktop
    RStudio Version :  1.1.456

R version 3.4.4 (2018-03-15)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.5

### Steps to reproduce the problem
1. Open .RProj from a mapped network drive folder (that you are connected to via VPN).  This doesn't seem to be a problem with files/projects stored on the local drive.  On Mac, e.g., Finder > Go > Connect to Server... Server Address = smb://<pathto>.RProj
2. Open a .R script file within that project folder.
3. Session -> New Session (as an aside: why this action opens the same project in a new session and doesn't just open a brand new, clean session really baffles me (feature request!), anyway...)
4. Go back to the original session and make an edit to the .R file.  A couple seconds later you get a pop-up: "Save File No such file or directory"  With an OK button.  No matter what you do from here on out you cannot save changes to the .R file.   

### Describe the problem in detail
Cannot save network drive R project script file after opening new session.

### Describe the behavior you expected
Successfully save the edited file.

Session Error info from diagnostic report:
(I've removed the actual network drive paths)

```
> 21 Sep 2018 15:52:24 [rsession-username] ERROR system error 45 (Operation not supported) [lock-file=/<pathto>/.Rproj.user/442D8A65/sources/s-30B2E4F4/lock_file]; OCCURRED AT: virtual rstudio::core::Error rstudio::core::AdvisoryFileLock::acquire(const rstudio::core::FilePath &) /Users/rstudio/rstudio/src/cpp/core/file_lock/AdvisoryFileLock.cpp:137; LOGGED FROM: bool rstudio::session::source_database::supervisor::<anonymous namespace>::reclaimOrphanedSession() /Users/rstudio/rstudio/src/cpp/session/SessionSourceDatabaseSupervisor.cpp:374

> 21 Sep 2018 15:52:24 [rsession-username] ERROR system error 45 (Operation not supported) [lock-file=/<pathto>/.Rproj.user/442D8A65/sources/s-30B2E4F4/lock_file]; OCCURRED AT: virtual rstudio::core::Error rstudio::core::AdvisoryFileLock::acquire(const rstudio::core::FilePath &) /Users/rstudio/rstudio/src/cpp/core/file_lock/AdvisoryFileLock.cpp:137; LOGGED FROM: rstudio::core::Error rstudio::session::source_database::supervisor::<anonymous namespace>::createSessionDir() /Users/rstudio/rstudio/src/cpp/session/SessionSourceDatabaseSupervisor.cpp:256

> 21 Sep 2018 15:53:54 [rsession-username] ERROR system error 77 (No locks available); OCCURRED AT: virtual rstudio::core::Error rstudio::core::AdvisoryFileLock::release() /Users/rstudio/rstudio/src/cpp/core/file_lock/AdvisoryFileLock.cpp:153; LOGGED FROM: rstudio::core::Error rstudio::session::source_database::supervisor::detachFromSourceDatabase() /Users/rstudio/rstudio/src/cpp/session/SessionSourceDatabaseSupervisor.cpp:566

> 21 Sep 2018 15:53:54 [rsession-username] ERROR system error 2 (No such file or directory); OCCURRED AT: rstudio::core::Error rstudio::core::FilePath::children(std::vector<FilePath> *) const /Users/rstudio/rstudio/src/cpp/core/FilePath.cpp:998; LOGGED FROM: rstudio::core::Error rstudio::session::source_database::supervisor::<anonymous namespace>::removeSessionDir(const rstudio::core::FilePath &) /Users/rstudio/rstudio/src/cpp/session/SessionSourceDatabaseSupervisor.cpp:159

```

### PR DESCRIPTION
**Pull request by kevinushey**
Closes #3521.

### CODE REVIEW

File: src/cpp/core/file_lock/FileLock.cpp [hunk 1]
```diff
@@ -209,7 +211,7 @@ void FileLock::log(const std::string& message)
 }
 
 // default values for static members
-FileLock::LockType FileLock::s_defaultType(FileLock::LOCKTYPE_LINKBASED);
+FileLock::LockType FileLock::s_defaultType(kLockTypeDefaultEnum);
```
Comment (jmcphers) [inline src/cpp/core/file_lock/FileLock.cpp]: If I'm reading this correctly, it not only makes link based locks possible, it also switches to link based locks by default on MacOS and Linux desktops. Is that intentional? (IIRC we don't have known issues with advisory locks there)

Comment (kevinushey) [reply src/cpp/core/file_lock/FileLock.cpp]: That's a good point -- we don't want to change the default here; just make it possible to respect the user-set lock for session locks. I'll fix this.

**Comment by jmcphers** [comment]
Won't that still change the lock type default on those operating systems?

https://github.com/rstudio/rstudio/blob/e9adaf7cf4b2fb9fb7090669b004b9cae8fcf5df/src/cpp/core/file_lock/FileLock.cpp#L212

**Comment by kevinushey** [comment]
Sorry about that! I think I've got the right implementation wired in now.

**Comment by jmcphers** [comment]
LGTM!

**Pull request merged**